xla/service/gpu/runtime/nccl_all_to_all_thunk.cc

```diff
@@ -145,13 +143,15 @@ absl::Status NcclAllToAllStartThunk::Cleanup(const CleanupParams& params) {
                         nccl_api()->CommCount(comm_wrapper.comm_handle));
 
     int local_id = params.executor->device_ordinal() % num_participants;
+    absl::MutexLock send_lock(&send_mutex_);
```

Comment:
It's better to do more granular locking. Please put this lock and the if below it in a nested scope:

```
{
   absl::MutexLock ...
    if ...
}
```

That will make sure the mutex is only locked for the block that needs it.